### PR TITLE
enhancement(observability): Track total allocations/deallocations

### PR DESCRIPTION
**Pull request by arshiyasolei**
Add support for tracking the accumulated memory deallocations and allocations per component.

### CODE REVIEW

**Review by tobz — CHANGES_REQUESTED** [review]
Just noted a small change that needs to be made but otherwise looks good to me. 👍🏻

File: src/internal_telemetry/allocations/mod.rs [hunk 1]
```diff
@@ -111,21 +118,45 @@ pub fn init_allocation_tracing() {
...
+                            allocations_diff.to_f64().expect("failed to convert allocations_diff from int to float"),
+                            "component_kind" => group_info.component_kind.clone(),
+                            "component_type" => group_info.component_type.clone(),
+                            "component_id" => group_info.component_id.clone());
+                        increment_gauge!(
+                            "component_deallocated_bytes_total",
+                            deallocations_diff.to_f64().expect("failed to convert deallocations_diff from int to float"),
+                            "component_kind" => group_info.component_kind.clone(),
+                            "component_type" => group_info.component_type.clone(),
+                            "component_id" => group_info.component_id.clone());
```
[the remark this answers is not in the document]
Comment (jszwedko) [reply src/internal_telemetry/allocations/mod.rs:144]: 👍 I was about to say the same thing, these seem like they should be counters.

[the remark this answers is not in the document]
Comment (arshiyasolei) [reply src/internal_telemetry/allocations/mod.rs:144]: There currently is an issue with counters resetting to zero when there's no reporting.

[the remark this answers is not in the document]
Comment (tobz) [reply src/internal_telemetry/allocations/mod.rs:144]: Counters don't display cumulatively on Datadog, so you should only expect to see spikes in the graphs when an update is emitted, which is consistent with what you're seeing.

**Review by tobz — APPROVED** [review]
Yeah, this makes sense to me.

Saw what looked like a small jump in the overhead to around -10% at worst, which is unfortunate... but I imagine we can claw that back with some profiling/data locality refactoring.

**Comment by arshiyasolei** [comment]
Yeah I think that 10% regression was due to not being updated w/ the main branch aha

**Pull request merged**